HIP/hip-1028.md

```diff
@@ -0,0 +1,157 @@
+
+## Specification
+
+We will utilize the existing Hedera Token Service (HTS) HAPI transactions.  
+To achieve this the Hedera Token Service (HTS) system contract must be updated to expose the necessary features.  
+Existing system contract functions must be extended and new functions must be added to the `IHederaTokenService` interface  
+to support the metadata field and key.
+
+### System contract extension
+
+The `IHederaTokenService.sol` is further extended with new types and function signatures that the system contract will respond to.
+
+### New Solidity Types
+
+The following structs have been added to support metadata and metadata key
+
+```solidity
+struct HederaTokenV2 {
+    String name;
+    String symbol;
+    // other fields are like in HederaToken
+    ...
+    // The new field for metadata
+    bytes metadata;
+}
+```
+
+```solidity
+struct TokenInfoV2 {
+    // The new struct with metadata
+    HederaTokenV2 token;
+    // other fields are like in TokenInfo
+    ...
+}
+```
+
+```solidity
+struct FungibleTokenInfoV2 {
+    // The new struct with metadata
+    TokenInfoV2 token;
+    int32 decimals;
+}
+```
+
+```solidity
+struct NonFungibleTokenInfoV2 {
+    // The new struct with metadata
+    TokenInfoV2 token;
+    // other fields are like in NonFungibleTokenInfo
+    ...
+}
+```
+Adding comment to TokenKey struct to clarify the 7th bit is for metadata key
+
+```solidity
+struct TokenKey {
+    // bit field representing the key type. Keys of all types that have corresponding bits set to 1
+    // will be created for the token.
+    // 0th bit: adminKey
+    // 1st bit: kycKey
+    // 2nd bit: freezeKey
+    // 3rd bit: wipeKey
+    // 4th bit: supplyKey
+    // 5th bit: feeScheduleKey
+    // 6th bit: pauseKey
+    
+    // the 7th bit will be metadata key
+    // 7th bit: metadataKey
+    uint keyType;
+    
+    // the value that will be set to the key type
+    KeyValue key
+}
+```
+
+### Solidity Function Signatures
+
+The new structs will change the signature of the functions. However the old versions of the functions will continue working.
+
+| Hash | Selector                                                                                                                                                                      | Return                                                                     |
+| --- |-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|----------------------------------------------------------------------------|
+|0fcaca1f| `updateNFTsMetadata(address token, int64[] memory serialNumbers, bytes memory metadata)`                                                                                      | `int responsecode`                                                         |
```

Comment:
Move to end to match order of use cases?  It's kind of the odd man out.  Another approach is to split this table into three lists (separate section headers or maybe not, just three separate tables: getters, setters, and then the update uniques method.